Mathlib/GroupTheory/GroupAction/FixedPoints.lean

```diff
@@ -0,0 +1,331 @@
+  Iff.rfl
+
+@[to_additive]
+theorem not_mem_movedBy_iff_mem_fixedBy {m : M} {a : α} : a ∉ movedBy α m ↔ a ∈ fixedBy α m := by
+  rw [movedBy_eq_compl_fixedBy, Set.not_mem_compl_iff]
+
+variable (α)
+
+/-- In a multiplicative group action, the points moved by `g` are also moved by `g⁻¹` -/
+@[to_additive "In an additive group action, the points moved by `g` are also moved by `g⁻¹`"]
+theorem movedBy_eq_movedBy_inv (g : G) : movedBy α g = movedBy α g⁻¹ := by
+  repeat rw [movedBy_eq_compl_fixedBy]
+  rw [fixedBy_eq_fixedBy_inv]
+
+variable {α}
+
+@[to_additive]
+theorem smul_mem_movedBy_iff_mem_movedBy {a : α} {g : G} :
+    g • a ∈ movedBy α g ↔ a ∈ movedBy α g := by
+  repeat rw [← not_mem_fixedBy_iff_mem_movedBy]
+  rw [smul_mem_fixedBy_iff_mem_fixedBy]
+
+@[to_additive]
+theorem smul_inv_mem_movedBy_iff_mem_movedBy {a : α} {g : G} :
+    g⁻¹ • a ∈ movedBy α g ↔ a ∈ movedBy α g :=
+  (movedBy_eq_movedBy_inv α g) ▸ smul_mem_movedBy_iff_mem_movedBy
+
+@[to_additive]
+theorem movedBy_zpow_subset_movedBy (g : G) (j : ℤ) :
+    movedBy α (g ^ j) ⊆ movedBy α g := by
+  repeat rw [movedBy_eq_compl_fixedBy]
+  rw [Set.compl_subset_compl]
+  exact fixedBy_subset_fixedBy_zpow α g j
+
+variable (M α) in
+@[to_additive (attr := simp)]
+theorem movedBy_one_eq_empty : movedBy α (1 : M) = ∅ := by
+  rw [movedBy_eq_compl_fixedBy, fixedBy_one_eq_univ, Set.compl_univ]
+
+variable (α) in
+@[to_additive]
+theorem movedBy_mul (m₁ m₂ : M) : movedBy α (m₁ * m₂) ⊆ movedBy α m₁ ∪ movedBy α m₂ := by
+  repeat rw [movedBy_eq_compl_fixedBy]
+  rw [← Set.compl_inter, Set.compl_subset_compl]
+  exact fixedBy_mul α m₁ m₂
+
+variable (α) in
+@[to_additive]
+theorem movedBy_conj (g h : G) :
+    movedBy α (h * g * h⁻¹) = (fun a => h⁻¹ • a) ⁻¹' movedBy α g := by
+  repeat rw [movedBy_eq_compl_fixedBy]
+  rw [Set.preimage_compl, fixedBy_conj]
+
+@[to_additive]
+theorem movedBy_conj' {g h : G} :
+    h • movedBy α g = movedBy α (h * g * h⁻¹) := by
+  rw [movedBy_conj, Set.preimage_smul, inv_inv]
+
+end MovedBy
+
+section Image
+
+/-!
+### Images of supersets of `movedBy α g`
+
+If a set `s : Set α` is a superset of `MulAction.movedBy α g` (resp. `AddAction.movedBy α g`),
+the no point or subset of `s` can be moved outside of `s` by the group action of `g`.
+-/
+
+/-- If `movedBy α g ⊆ s`, then `g` cannot move a point of `s` outside of `s`. -/
+@[to_additive "If `movedBy α g ⊆ s`, then `g` cannot move a point of `s` outside of `s`."]
+theorem smul_zpow_mem_of_movedBy_subset {a : α} {s : Set α} {g : G} (s_subset : movedBy α g ⊆ s)
+    (j : ℤ) : g ^ j • a ∈ s ↔ a ∈ s := by
+  by_cases a ∈ movedBy α (g ^ j)
+  case pos a_moved =>
+    constructor
+    all_goals (intro; apply s_subset)
+    all_goals apply movedBy_zpow_subset_movedBy g j
+    · exact a_moved
+    · rw [smul_mem_movedBy_iff_mem_movedBy]
+      exact a_moved
+  case neg a_fixed =>
+    rw [not_mem_movedBy_iff_mem_fixedBy, mem_fixedBy] at a_fixed
+    rw [a_fixed]
```

Comment:
Here's an alternate proof, not because I'm suggesting you adopt it, but because it indicates to me that the "substance" of this lemma is `g • s = s` (`g` takes `s` to `s`). So I wonder whether some of your later results might be better phrased in this pointwise manner. For instance, `smul_zpow_mem_fixedBy_of_commute` could be phrased as `Commute g h → h • fixedBy α g = fixedBy α g`.

```
have h1 : g • s = s
  · ext a
    by_cases h : g • a = a
    · rw [Set.mem_smul_set_iff_inv_smul_mem, ← h, inv_smul_smul, h]
    rw [Set.mem_smul_set_iff_inv_smul_mem]
    exact iff_of_true (s_subset (smul_inv_mem_movedBy_iff_mem_movedBy.mpr h)) (s_subset h)
  have h2 : g ^ j • s = s
  · exact fixedBy_subset_fixedBy_zpow (Set α) g j h1
  rw [← Set.ext_iff.mp h2 (g ^ j • a), Set.smul_mem_smul_set_iff]
```

Reply:
You are right, if I go with the pointwise smul route, then I can re-use a lot of the theorems from above (since `g • s = s` is definitionally equal to `s ∈ fixedBy (Set α) g`). I was wary of using pointwise smul, because they are used rarely enough that I didn't know about it for a while at first.